wgsl/index.bs

```diff
@@ -513,7 +514,205 @@ When unclear from context, this specification indicates
+6. If processing during [=shader module creation=] time, the remaining diagnostics populate the
+    {{GPUCompilationInfo/messages}} member of the WebGPU {{GPUCompilationInfo}} object.
+7. If processing during [=pipeline creation=], the [=severity/error=] diagnostics are reported
+    in the nearest enclosing WebGPU [=GPU error scope=].
+    Issue: TODO: Adjust for [#3682](https://github.com/gpuweb/gpuweb/issues/3682).
+
+Note: The rules allow an implementation to stop processing a WGSL program as soon as an error is detected.
+Additionally, an analysis for a particular kind of warning can stop on the first warning,
+and an analysis for a particular kind of info diagnostic can stop on the first occurrence.
+WGSL does not specify the order to perform different kinds of analyses, or an ordering within a single analysis.
+Therefore, for the same WGSL program, different implementations may report different instances of diagnostics with the same severity.
+
+### Filterable Triggering Rules ### {#filterable-triggering-rules}
+
+Most [=diagnostics=] are unconditionally reported to the WebGPU application.
+Some kinds of diagnostics can be [[#diagnostic-filtering|filtered]], in part by naming their [=diagnostic/triggering rule=].
+The following table lists the standard set of triggering rules that can be filtered.
+
+<table class='data'>
+  <caption>Filterable diagnostic triggering rules</caption>
+  <thead>
+    <tr><th>Filterable Triggering Rule<th>Default Severity<th>Triggering Location<th>Description
+  </thead>
+
+  <tr>
+    <td><dfn noexport dfn-for="trigger">derivative_uniformity</dfn>
+    <td>[=severity/error=]
+    <td>The [=call site=] for any of the
+        [=builtin functions that compute a derivative|builtin function that computes a derivative=].
+        That is, the [=call site=] for any of:
+        * the [[#derivative-builtin-functions|derivative builtin functions]]
+        * [[#texturesample|textureSample]]
+        * [[#texturesamplebias|textureSampleBias]]
+        * [[#texturesamplecompare|textureSampleCompare]]
+
+    <td>A call to a builtin function computes derivatives, but [=uniformity analysis=] cannot prove that the call occurs in [=uniform control flow=].
+
+    See [[#uniformity]].
+</table>
+
+An implementation may support filtering triggering rules not specified here,
+provided they are spelled differently from the standard triggering rules.
+A [=diagnostic filter=] with an unrecognized [=diagnostic/triggering rule=] [=behavioral requirement|will=] be ignored.
+
+Future versions of this specification may remove a particular rule or weaken its default severity
+(i.e. replace its current default with a less severe default) and still be deemed as satisfying backward compatibility.
+For example, a future version of WGSL may change the default severity for [=trigger/derivative_uniformity=] from `error` to either `warning` or `info`.
+After such a change to the specification, previously valid programs would remain valid.
```

Comment:
I don't know how much it matters, but I can image scenarios where this might not quite be true. Consider an implementation that produces a warning for an unrecognized diagnostic rule (as Tint does currently), that also provides an implementation-specific rule for controlling the severity of unrecognized diagnostic rules. A shader author might want unrecognized rules to be hard errors, so they write this:
```
diagnostic(error, unrecognized_diagnostic_rule);
diagnostic(off, derivative_uniformity);
```
Now, if we were to remove the `derivative_uniformity` rule from WGSL in the future, this shader would start producing a shader-creation error when it didn't previously.

Maybe this is a corner case that we are OK with? Or, we have to guarantee that we will never remove a rule from the spec (but certain rules may become obsolete instead).

Reply:
Sigh.  When first writing this up, I didn't think of the idea of "unrecognized_diagnostic_rule". 

You're right, we could have a an explicit list of obsoleted rules. They are always "off" in some sense to be worked out. ?